### PR TITLE
WIP - Set up bats based test bed

### PR DESCRIPTION
**Pull request by felixSchl**
Complementary PR for #469 

I will remove tabs in favour spaces for consistency once I am done or closing in on being done. I find it quite pleasant to work with hard tabs in bash.

### CODE REVIEW

**Comment by felixSchl** [comment]
I think the it's at a stage where some initial feedback would be helpful. To explain what I have done so far:

The test bed itself is shell agnostic, but depending on how you execute bats, a different shell will be used to interact with liquidprompt:

```
TEST_SHELL=bash bats/tests
TEST_SHELL=zsh bats/tests
```

Most of this logic sits in [support.bash](https://github.com/felixSchl/liquidprompt/blob/f4339f713e46b1ec4ae47a80bc6cbdfdc3ff603d/tests/support.bash).

I think that's the right way to do it because liquidprompt aims to work out of the box on each. We can specialize tests by checking `$TEST_SHELL`, if need be.

I have also [implemented a function that allows fetching symbols from the underlying shell](https://github.com/felixSchl/liquidprompt/blob/f4339f713e46b1ec4ae47a80bc6cbdfdc3ff603d/tests/support.bash#L122), so that they can be compared in the host shell. This allows to check for a symbol in PS1 irrespective of the host shell. You could probably do more with `eval` here in terms of implementation, but I I think that would add too much complexity.

_What I do **not** like this set up_, is that currently the code interacting with liquidprompt and the code evaluating the results is [separate](https://github.com/felixSchl/liquidprompt/blob/f4339f713e46b1ec4ae47a80bc6cbdfdc3ff603d/tests/liquidprompt.bats#L31-L129), creating a disjoint. What would be nice if we could open a interactive shell in the background and send commands to it using fifo or something.

_What I **do** like about this set up_, is that's very close to being an actual pristine environment (afaict) for every single test case, for every single `run` call even.

**Comment by dolmen** [comment]
Personally I also prefer hard tabs for shell scripting. But my experience as a maintainer when working with multiple casual contributors is that when using tabs you have to spend too much time rejecting patches that have spaces instead of tabs. And using a different indent style in the testsuite from the main code would make the problem even worse.

So the testsuite of liquidprompt should use the same indent style as the liquidprompt source: 4-spaces.

I'll review your PR later.

**Comment by felixSchl** [comment]
Cool. There's currently a lot of noise again, trying to get things to go on travis.yml. I agree with respect to consistency and I will switch it over very soon, before the history would get too messy but one thing should be mentioned: it seems that shell was supposed to be written with hard tabs because there's special support for heredocs, allowing me to indent them rather than glueing them to the left of the file with `<<-EOF` which is really handy. I guess I could still use tabs leading up to the block and then drop them, but a bit of a hassle to work with (in vim at least).

**Comment by felixSchl** [comment]
I drastically simplified the approach and removed plenty of code. Rather than using heredoc, the code a single test lives now in a file in `tests/cases/*`. There's a bats file that would literally just do `@test 'lp: git' { run_shell tests/cases/git.sh }`. The file extension is `.sh` just to help file editors pick it up.

Some notes:

* Travis is set up for my fork, and can be seen here: https://travis-ci.org/felixSchl/liquidprompt
* Tests are run w/o colour atm, and the entire PS1 can be checked for equality this way. We can invest more time into finding a way test colour also.
* There are mocks living in `tests/support/mocked-liquidprompt.sh` - some are defined prior to liquidprompt start, others after. The ones defined prior are required for feature detection or where liquidprompt caches stuff right from the get go, such as the connection type. (note: i have had to make a slight change to liquidprompt: https://github.com/felixSchl/liquidprompt/blob/f88f3fa368b716b00924c19020614cb2507efb24/liquidprompt#L587 to enable this).
* The subshell is launched using `env -i` and only select variables are assigned for it, providing further isolation
* `eval_prompt` calls `_lp_set_prompt` but also rewires the assertion functions so the assertions following are specific to the changes made since the last `eval_prompt` call or since the beginning of the script.
* For the git tests, I am using real git, which means the tests runs a bit slower but we know it works against real git. we could provide mocks here, too, if you think that's better.

I thought we could also generate the bats file on the fly, but maybe a bit too early for that.

Not entirely sure where from here, I think we could just start adding tests

**Comment by felixSchl** [comment]
I've gone ahead and added some path shorten tests, and I found this interesting. Probably this thread is not the place to discuss the ins and outs, but briefly:

```
for t in \
    {0..8}:'/tmp … path' \
    9:'/tmp … /path' \
    10:'/tmp … g/path'
do
    IFS=: read -r ns p <<< "$t"
    for n in $ns; do
        LP_PATH_LENGTH="$n"
        source "$SCRIPT_DIR/tests/support/mocked-liquidprompt.sh"
        eval_prompt "LP_PATH_LENGTH=$n -> $p"
        assert_path "$p"
    done
done
```

will pass on bash, but cause a failure on zsh:

```
   ...check: LP_PATH_LENGTH=9 -> /tmp … /path
   Expected PS1 to equal "0.64 4d [\u:/tmp … /path] 2s \$ " - got: "0.64 4d [\u:/tmp … path] 2s \$ "
   ...check: LP_PATH_LENGTH=10 -> /tmp … g/path
   Expected PS1 to equal "0.64 4d [\u:/tmp … g/path] 2s \$ " - got: "0.64 4d [\u:/tmp … path] 2s \$ "
```

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  - TEST_CONNECTION=lcl TEST_SHELL=zsh
+
+before_install:
+  - |
+    if [[ "$TRAVIS_OS_NAME" == osx ]]; then
+      brew install bats zsh
+    else
+      sudo add-apt-repository ppa:duggan/bats --yes
+      sudo apt-get update -qq
+      sudo apt-get install -qq -y bats zsh
```
Comment (dolmen) [inline .travis.yml]: Use `addons.apt.packages` instead.
https://docs.travis-ci.com/user/installing-dependencies/#Installing-Packages-on-Container-Based-Infrastructure

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+*.un~
+*.sw[a-z]
```
Comment (dolmen) [inline .gitignore]: `.sw?` is for Vim files. Adding a `.gitignore` for those files is out of scope for this PR.
I recommend you to instead have a global gitignore (see the core.excludesfile option of git).

What about `*.un~`?

File: test.sh [hunk 3]
```diff
@@ -6,9 +6,9 @@ if [ -z "$BASH_VERSION$ZSH_VERSION" ]; then
     for sh in bash zsh
     do
 	if [ -x /bin/$sh ]; then
-	    /bin/$sh "$0"
+		/bin/$sh "$0"
 	elif [ -x /usr/bin/$sh ]; then
-	    /usr/bin/$sh "$0"
+		/usr/bin/$sh "$0"
 	fi
```
Comment (dolmen) [inline test.sh]: Just use bash/zsh from $PATH: `$sh "$0"`

File: tests/support.bash [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+esac
+
+function setup
+{
+    export TMP_DIR="$BATS_TMPDIR/liquidprompt-tests"
+    rm -rf "$TMP_DIR"
+    mkdir -p "$TMP_DIR"
+
+    # on OSX, we are handed a deep path to some
+    # /var/private/tmp/ directory
```
Comment (dolmen) [inline tests/support.bash]: So, why is it a problem?

File: liquidprompt [hunk 5]
```diff
@@ -584,6 +584,7 @@ fi
 # Where are we? #
 #################
 
+if ! type _lp_connection &> /dev/null; then
```
Comment (dolmen) [inline liquidprompt]: This breaks my usual developement workflow: this check stops from updating liquidprompt's code by just reloading it in the same shell.
We should find a better way.

File: tests/support/assert.sh [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+function eval_prompt
+{
+    if [[ -n "$1" ]]; then
+        echo "...check: $1"
+    fi
+
+    export _LP_PS1 _LP_PS1_PLAIN
+    _lp_set_prompt
+    _LP_PS1="$PS1"
+    _LP_PS1_PLAIN="$(strip_colors <<< "$_LP_PS1")"
```
Comment (dolmen) [inline tests/support/assert.sh]: Don't use `_LP_` prefix (and neither `LP_`) for variables in the testsuite but not in liquidprompt.

File: tests/support/mocked-liquidprompt.sh [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        for i in eval "echo {1..$DETACHED_JOB_COUNT_SCREEN}"; do
+            echo 'Detach)'
+        done
+    else
+        "$_screen" "$@"
+    fi
+}
+
+
+source "$SCRIPT_DIR/liquidprompt"
```
Comment (dolmen) [inline tests/support/mocked-liquidprompt.sh]: A fixed liquidprompt configuration file should be used.
Currently this testsuite fails on my environment because I have a `~/.liquidpromptrc` with non default settings.

File: tests/support/mocked-liquidprompt.sh [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+DETACHED_SESSION_COUNT_TMUX=2
+DETACHED_SESSION_COUNT_SCREEN=2
+DETACHED_SESSION_COUNT=4
+
+_tmux="$(which tmux)"
+function tmux
+{
+    if [[ "$1" == list-sessions ]]; then
+        for i in eval "echo {1..$DETACHED_JOB_COUNT_TMUX}"; do
```
Comment (dolmen) [inline tests/support/mocked-liquidprompt.sh]: - `i` must be local
- The loop doesn't work:
```
$ for i in eval "echo {1..$x}"; do
> echo $i
> done
eval
echo {1..4}
```

File: tests/support/mocked-liquidprompt.sh [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    else
+        "$_tmux" "$@"
+    fi
+}
+
+_screen="$(which screen)"
+function screen
+{
+    if [[ "$1" == -ls ]]; then
+        for i in eval "echo {1..$DETACHED_JOB_COUNT_SCREEN}"; do
```
Comment (dolmen) [inline tests/support/mocked-liquidprompt.sh]: - `i` must be local
- The loop doesn't work

File: tests/support/mocked-liquidprompt.sh (hunk 7 above)
Comment (felixSchl) [inline tests/support/mocked-liquidprompt.sh]: I think I can achieve this by re-setting "$HOME" during the test run

File: liquidprompt (hunk 5 above)
Comment (felixSchl) [inline liquidprompt]: we could add a check like `! type _mock_lp_connection` and then make `_lp_connection` call that. But that would introduce test specific code...

File: tests/support/mocked-liquidprompt.sh (hunk 8 above)
Comment (felixSchl) [inline tests/support/mocked-liquidprompt.sh]: thanks, fixed in later commits (also the one below)

File: test.sh (hunk 3 above)
Comment (felixSchl) [inline test.sh]: Sorry, this was not mine, i just accidentally changed indentation. I reverted my change.

File: .gitignore (hunk 2 above)
Comment (felixSchl) [inline .gitignore]: Removed from repo

File: tests/support.bash (hunk 4 above)
Comment (felixSchl) [inline tests/support.bash]: Because path shortening would yield different results based on platform. I will add to the comment

File: .travis.yml (hunk 1 above)
Comment (felixSchl) [inline .travis.yml]: I've done this, but it appears to be a beta feature on travis and travis is not happy: https://travis-ci.org/felixSchl/liquidprompt/jobs/162850367

**Comment by nojhan** [comment]
This PR is old but a test suite would be very interesting, can I consider merging it or is there something else to change since two years ago?

**Comment by Rycieos** [comment]
I'm closing this PR as the tests are not compatible with the reworked code. 

The point of the rework was to separate data gathering from data display (data vs theme). This had the side effect of making testing the data sources much simpler, which I have done (for most of the data sources) with the new test suite.

As I see it, this PR adds testing for:
* Git
* Path shortening
* Prompt tags (PS1_PREFIX)
* Digital time
* Analog time

None of which (besides Git and somewhat analog time) are data sources. I would be fine with testing on the theme layer, but it would have to be done by mocking the new data source layer. And with the split of data from theme, most of the theme functions are so simple, any test built for them would be exactly duplicating the function they are testing, which defeats the point. (Ex, `_lp_username_color()`. But `_lp_hostname_color()` should probably be tested.)

As for Git (and other VCS providers) I would like to see testing on them, but they can be a bit tricky to get right (as this PR shows). If you fixup the Git tests for the new test suite, I would merge them, but writing VCS tests is not the top priority for me right now.

So thanks for the work you put into this, but unfortunately the state of the project when you wrote this just wasn't good enough to make testing work well.

If you wanted to write some tests for the new test suite, I would be happy to merge them.
